### ISSUE TITLE
Support Closure as a JavaScript minifier option

### ISSUE THREAD

**Issue by matthewp** [stealjs/steal-tools#503]
Closure has good [es6 support](https://github.com/google/closure-compiler/wiki/ECMAScript6), better than Uglify.  Maybe we can make the `minify` option take a string:

```js
{
  minify: true // Default, use uglify
}
```

```js
{
  minify: false // Don't minify
}
```

```js
{
  minify: 'closure' // Use closure
}
```

**Comment by matthewp** [comment]
This is the npm package to use I think: https://www.npmjs.com/package/google-closure-compiler

**Comment by m-mujica** [comment]
@matthewp I was looking into this last night, if I'm not mistaken we have to use this API https://github.com/chadkillingsworth/closure-compiler-npm#plugin-authors-and-native-node-usage but I noticed it's async (I guess the minified code is the `stdOut` parameter), with UglifyJS the `minify` function in sync. 

Should we make this https://github.com/stealjs/steal-tools/blob/master/lib/buildTypes/minifyJS.js#L4 return a promise that resolves to the minified code? I'm not sure how big of a change it would be.

**Comment by matthewp** [comment]
Yep! That definitely sounds right to me.

I don't think it's too big of a change, the minification happens as it's own stream, here: https://github.com/stealjs/steal-tools/blob/2f945de2d434d3b8cfbce6dea1c7d7b8c2a096c8/lib/stream/minify.js

I think you just make it return a Promise like you said, then wait on those promises before resolving the stream.

**Comment by m-mujica** [comment]
@matthewp, the closure compile lib takes an array of file paths, it does not seem to support natively string source code. I found this wrapper https://github.com/simone-sanfratello/node-closure-compiler-wrapper which I think is creating just a temp file when given a string. 

Any suggestion? should we create temp files with the source code, too?  We should have the location of the file in the load data, but I guess that won't help for files that need to be transpiled first.

**Comment by matthewp** [comment]
I would do the tmp stuff yourself, that lib is not as up to date as google's official one. I've used [tmp](https://www.npmjs.com/package/tmp) in the past, so it's pretty easy.

That's too bad, maybe in 1.0 we should only minify the full bundle after concatenation (I think we do this for css already) and not the individual modules. Or at least by default, maybe leave in the other way for slightly better debugging.

**Comment by m-mujica** [comment]
@matthewp I almost done with this issue https://github.com/stealjs/steal-tools/compare/support-closure?expand=1, but I have some little problems :)

Closure compiler uses some ES6 features, tests won't for older versions of Node.

Should we drop this issue and instead make steal use a `minify` function and let uglify as the default instead? Should we close this one or just update the title?

**Comment by matthewp** [comment]
A new issue is probably best, just reference this old one so there's a paper trail.

**Comment by pYr0x** [comment]
@m-mujica maybe merge that into the `major` (steal 1.0) branch?

**Comment by m-mujica** [comment]
@pYr0x we are going that discard that branch and try a different approach.... Instead of supporting every possible minification library out there, we are going to make it possible for steal-tools to take a `minify` function. That way we don't have to bundle closure (or any other minification library besides the default, which will be UglifyJS). Makes sense?

**Comment by pYr0x** [comment]
@m-mujica can we use this not already with your stream API?

**Comment by matthewp** [comment]
The stream API is a little too advanced for most cases though.

Another thing we are doing here is moving minification to the bundle level instead of the module level so this will be a breaking change. I guess the minify stream will need to change what it currently does.
